it/tutorials/scala-for-java-programmers.md

```diff
@@ -0,0 +1,716 @@
+necessario ricordare le seguenti regole che riguardano questa
+operazione:
+
+1.  la derivata di una somma è la somma delle derivate,
+2.  la derivata di una variabile `v` è uno se `v` è la variabile di
+    derivazione, zero altrimenti,
+3.  la derivata di una costante è zero.
+
+Queste regole possono essere tradotte quasi letteralmente in codice Scala e
+ottenere la seguente definizione:
+
+    def derive(t: Tree, v: String): Tree = t match {
+      case Sum(l, r) => Sum(derive(l, v), derive(r, v))
+      case Var(n) if (v == n) => Const(1)
+      case _ => Const(0)
+    }
+
+Questa funzione introduce due nuovi concetti relativi al pattern
+matching. Prima di tutto l’istruzione `case` per le variabili ha un
+*controllo*, un’espressione che segue la parola chiave `if`. Questo
+controllo fa si che il pattern matching è eseguito solo se l’espressione
+è vera. Qui viene usato per esser sicuri che restituiamo la costante `1`
+solo se il nome della variabile da derivare è lo stesso della variabile
+di derivazione `v`. La seconda nuova caratteristica del pattern matching qui
+introdotta è la *wild-card*, scritta `_`, che corrisponde a qualunque
+valore, senza assegnargli un nome.
+
+Non abbiamo esplorato del tutto la potenza del pattern matching ma ci
+fermiamo qui per brevità. Vogliamo ancora osservare come le due
+precedenti funzioni lavorano in un esempio reale. A tale scopo
+scriviamo una semplice funzione `main` che esegue diverse operazioni
+sull’espressione `(x+x)+(7+y)`: prima calcola il suo valore
+nell’environment `{ x -> 5, y -> 7 }`, dopo calcola la
+derivata relativa ad `x` e poi ad `y`.
+
+    def main(args: Array[String]) {
+      val exp: Tree = Sum(Sum(Var("x"),Var("x")),Sum(Const(7),Var("y")))
+      val env: Environment = { case "x" => 5 case "y" => 7 }
+      println("Expression: " + exp)
+      println("Evaluation with x=5, y=7: " + eval(exp, env))
+      println("Derivative relative to x:\n " + derive(exp, "x"))
+      println("Derivative relative to y:\n " + derive(exp, "y"))
+    }
+
+Eseguendo questo programma otteniamo l’output atteso:
+
+    Expression: Sum(Sum(Var(x),Var(x)),Sum(Const(7),Var(y)))
+    Evaluation with x=5, y=7: 24
+    Derivative relative a x:
+     Sum(Sum(Const(1),Const(1)),Sum(Const(0),Const(0)))
+    Derivative relative to y:
+     Sum(Sum(Const(0),Const(0)),Sum(Const(0),Const(1)))
+
+Esaminando l’output notiamo che il risultato della derivata dovrebbe
+essere semplificato prima di essere visualizzato all’utente. La
+definizione di una funzione di semplificazione usando il pattern
+matching rappresenta un interessante (ma sorprendentemente
+ingannevole) problema  che lasciamo come esercizio per il lettore.
+
+
+## I Trait
+
+Una classe in Scala oltre che poter ereditare da una super-classe può anche
+importare del codice da uno o più *trait*.
+
+Probabilmente per i programmatori Java il modo più semplice per capire cosa
+sono i trait è concepirli come interfacce che possono contenere del codice.
```

Comment:
I think for someone with Java background they are closer to an abstract class (but one that can't have a constructor). Also, now that Java8 has default methods, the one important different between Scala trait and Java8 interfaces is that trait can also contain state, and not just behavior.